packages/preact/test/test.js

```diff
@@ -1,107 +1,224 @@
-const CustomDelete = props => <del style={{color: 'crimson'}} {...props} />
-const Layout = ({children}) => <div id="layout">{children}</div>
-
-it('allows components to be passed via context', () => {
-  const result = render(
-    <MDXProvider components={{h1: H1, wrapper: Layout}}>
-      <Fixture />
-    </MDXProvider>
-  )
-
-  // Layout is rendered
-  expect(result).toMatch(/id="layout"/)
-
-  // H1 is rendered
-  expect(result).toMatch(/style="color: tomato;"/)
+import {transformAsync as babelTransform} from '@babel/core'
+import mdxTransform from '../../mdx'
+import {MDXProvider, withMDXComponents, mdx} from '../src'
+
+const run = async value => {
+  // Turn the serialized MDX code into serialized JSX…
+  const doc = await mdxTransform(value, {skipExport: true})
+
+  // …and that into serialized JS.
+  const {code} = await babelTransform(doc, {
+    configFile: false,
+    plugins: [
+      '@babel/plugin-transform-react-jsx',
+      'babel-plugin-remove-export-keywords'
+    ]
+  })
+
+  // …and finally run it, returning the component.
+  // eslint-disable-next-line no-new-func
+  return new Function('mdx', `${code}; return MDXContent`)(mdx)
+}
+
+describe('@mdx-js/preact', () => {
+  test('should evaluate MDX code', async () => {
+    const Content = await run('# hi')
+
+    expect(render(<Content />)).toEqual('<h1>hi</h1>')
+  })
+
+  test('should evaluate some more complex MDX code (text, inline)', async () => {
+    const Content = await run(
+      '*a* **b** `c` <abbr title="Markdown + JSX">MDX</abbr>'
+    )
+
+    expect(render(<Content />)).toEqual(
+      '<p><em>a</em> <strong>b</strong> <code>c</code> <abbr title="Markdown + JSX">MDX</abbr></p>'
+    )
+  })
+
+  test('should evaluate some more complex MDX code (flow, block)', async () => {
+    const Content = await run('***\n> * 1. a')
+
+    expect(render(<Content />)).toEqual(
+      '<hr /><blockquote><ul><li><ol><li>a</li></ol></li></ul></blockquote>'
+    )
+  })
+
+  test('should warn on missing components', async () => {
+    const Content = await run('<Component />')
+    const warn = console.warn
+    console.warn = jest.fn()
+
+    expect(render(<Content />)).toEqual('<div></div>')
+
+    expect(console.warn).toHaveBeenCalledWith(
+      'Component Component was not imported, exported, or provided by MDXProvider as global scope'
+    )
+
+    console.warn = warn
+  })
+
+  test('should support components defined in MDX', async () => {
+    const Content = await run('export const A = () => <b>!</b>\n\n<A />')
+
+    expect(render(<Content />)).toEqual('<b>!</b>')
+  })
+
+  test('should not crash if weird values could come from JSX', async () => {
+    // As JSX is function calls, that function can also be used directly in
+    // MDX. Definitely not a great idea, but it’s an easy way to pass in funky
+    // values.
+    const Content = await run('{mdx(1)}')
+
+    expect(render(<Content />)).toEqual('<1></1>')
+  })
```

Comment:
How far should this idea be taken?
Should imcomplete JSX crash or not crash? `<a>test</b>`
What about kinda, but not quite import statements? `import Something from './somewhere"`

Reply:
The idea should not go anywhere: there was code handling invalid values, which can’t be passed by JSX, so I couldn’t test it. I didn’t want to remove the code because I may be missing something, and if it can be removed, I’d want to do it in a different PR

Reply:
And: mdx-js/preact, mdx-js/react, and mdx-js/vue are runtime projects, whereas broken JSX or import/exports are compile time problems (in mdx-js/mdx)